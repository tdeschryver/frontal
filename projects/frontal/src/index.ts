export * from './public_api';
export { updateState } from '../src/lib/actions';
