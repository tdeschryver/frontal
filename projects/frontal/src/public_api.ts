/*
 * Public API Surface of frontal
 */

export {
  UpdateState,
  UpdateItems,
  SetItem,
  StateChanges,
  Action,
  ListToggle,
  InputChange,
  ListOpen,
  ListClose,
  ButtonClick,
  InputBlur,
  ItemMouseClick,
  ItemMouseEnter,
  ItemMouseLeave,
  InputKeydownArrowDown,
  InputKeydownArrowUp,
  InputKeydownEnter,
  InputKeydownEsc,
} from './lib/actions';
export { State, initialState, createState } from './lib/state';
export { FrontalModule } from './lib/frontal.module';
export {
  FrontalButtonDirective,
  FrontalComponent,
  FrontalInputDirective,
  FrontalLabelDirective,
  FrontalItemDirective,
  FrontalListDirective,
  FRONTAL_VALUE_ACCESSOR,
} from './lib/frontal.component';
export { StatusMessagePipe } from './lib/status.pipe';
export { generateId, resetId } from './lib/utils';
