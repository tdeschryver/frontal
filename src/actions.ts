export enum StateChanges {
  ListToggle = 'frontal_list_toggle',
  ListOpen = 'frontal_list_open',
  ListClose = 'frontal_list_close',
  InputFocus = 'frontal_input_focus',
  InputBlur = 'frontal_input_blur',
  InputChange = 'frontal_input_change',
  InputKeydownArrowDown = 'frontal_input_keydown_arrow_down',
  InputKeydownArrowUp = 'frontal_input_keydown_arrow_up',
  InputKeydownEnter = 'frontal_input_keydown_enter',
  InputKeydownEsc = 'frontal_input_keydown_esc',
  ItemMouseClick = 'frontal_item_mouseclick',
  ItemMouseEnter = 'frontal_item_mouseenter',
  ItemMouseLeave = 'frontal_item_mouseleave',
  ButtonClick = 'frontal_button_click',
}

export interface MenuToggle {
  type: StateChanges.ListToggle;
  payload: {
    isOpen: boolean;
  };
}

export interface InputFocus {
  type: StateChanges.InputFocus;
  payload: {};
}

export interface InputChange {
  type: StateChanges.InputChange;
  payload: {
    highlightedIndex: number | null;
    selectedItem: any;
    inputText: string;
    inputValue: string;
    isOpen: boolean;
  };
}

export interface MenuOpen {
  type: StateChanges.ListOpen;
  payload: {
    isOpen: boolean;
  };
}

export interface MenuClose {
  type: StateChanges.ListClose;
  payload: {
    isOpen: boolean;
  };
}

export interface ButtonClick {
  type: StateChanges.ButtonClick;
  payload: {
    isOpen: boolean;
  };
}

export interface InputBlur {
  type: StateChanges.InputBlur;
  payload: {
    isOpen: boolean;
    highlightedIndex: null;
    selectedItem: any;
    inputText: string;
    inputValue: string;
  };
}

export interface ItemMouseClick {
  type: StateChanges.ItemMouseClick;
  payload: {
    isOpen: boolean;
    highlightedIndex: null;
    selectedItem: any;
    inputText: string;
    inputValue: string;
  };
}

export interface ItemMouseEnter {
  type: StateChanges.ItemMouseEnter;
  payload: {
    highlightedIndex: number;
  };
}

export interface ItemMouseLeave {
  type: StateChanges.ItemMouseLeave;
  payload: {
    highlightedIndex: null;
  };
}

export interface InputKeydownArrowDown {
  type: StateChanges.InputKeydownArrowDown;
  payload: {
    selectedItem: null;
    highlightedIndex: number | null;
  };
}

export interface InputKeydownArrowUp {
  type: StateChanges.InputKeydownArrowUp;
  payload: {
    selectedItem: null;
    highlightedIndex: number | null;
  };
}

export interface InputKeydownEnter {
  type: StateChanges.InputKeydownEnter;
  payload: {
    isOpen: boolean;
    highlightedIndex: number | null;
    selectedItem: any;
    inputText: string;
    inputValue: string;
  };
}

export interface InputKeydownEsc {
  type: StateChanges.InputKeydownEsc;
  payload: {
    isOpen: boolean;
    highlightedIndex: number | null;
    selectedItem: any | null;
    inputText: string;
    inputValue: string;
  };
}

export type Action =
  | MenuToggle
  | InputChange
  | MenuOpen
  | MenuClose
  | ButtonClick
  | InputFocus
  | InputBlur
  | ItemMouseClick
  | ItemMouseEnter
  | ItemMouseLeave
  | InputKeydownArrowDown
  | InputKeydownArrowUp
  | InputKeydownEnter
  | InputKeydownEsc;
