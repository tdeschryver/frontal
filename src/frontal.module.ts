import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import {
  FrontalComponent,
  FrontalInputDirective,
  FrontalLabelDirective,
  FrontalItemDirective,
  FrontalButtonDirective,
  FrontalListDirective,
} from './frontal.component';

const components = [
  FrontalComponent,
  FrontalInputDirective,
  FrontalLabelDirective,
  FrontalItemDirective,
  FrontalButtonDirective,
  FrontalListDirective,
];

@NgModule({
  imports: [CommonModule],
  declarations: components,
  exports: components,
})
export class FrontalModule {}
