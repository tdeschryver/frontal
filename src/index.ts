export { FrontalModule } from './frontal.module';
export {
  StateChanges,
  Action,
  MenuToggle,
  InputChange,
  MenuOpen,
  MenuClose,
  ButtonClick,
  InputBlur,
  ItemMouseClick,
  ItemMouseEnter,
  ItemMouseLeave,
  InputKeydownArrowDown,
  InputKeydownArrowUp,
  InputKeydownEnter,
  InputKeydownEsc,
} from './actions';
export { State, initialState, createState } from './state';
export {
  FrontalButtonDirective,
  FrontalComponent,
  FrontalInputDirective,
  FrontalLabelDirective,
  FrontalItemDirective,
  FrontalListDirective,
  FRONTAL_VALUE_ACCESSOR,
} from './frontal.component';
