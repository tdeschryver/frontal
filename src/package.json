{
  "name": "frontal",
  "version": "1.0.0",
  "description": "An Angular select/dropdown component",
  "module": "frontal.js",
  "main": "bundles/frontal.umd.js",
  "typings": "frontal.d.ts",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/tdeschryver/frontal.git"
  },
  "keywords": [
    "Angular",
    "select",
    "selectbox",
    "combobox",
    "typeahead",
    "autocomplete",
    "dropdown"
  ],
  "author": "tdeschryver",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/tdeschryver/frontal/issues"
  },
  "homepage": "https://github.com/tdeschryver/frontal#readme",
  "peerDependencies": {
    "@angular/common": "^6.0.0-rc.3",
    "@angular/core": "^6.0.0-rc.3",
    "@angular/forms": "^6.0.0-rc.3"
  }
}
