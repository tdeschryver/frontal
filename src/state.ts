import { Action } from './actions';
import { FrontalItemDirective } from './frontal.component';
import { generateId } from './utils';

export interface State {
  id: string;
  selectedItem: any | null;
  highlightedIndex: number | null;
  defaultHighlightedIndex: number | null;
  highlightedItem: any | null;
  inputValue: string;
  inputText: string;
  isOpen: boolean;
  itemCount: number;
  reducer: (state: State, action: Action) => Action;
  itemToString: (value: any) => string;
}

export const initialState: State = {
  id: '',
  selectedItem: null,
  highlightedIndex: null,
  defaultHighlightedIndex: null,
  highlightedItem: null,
  inputValue: '',
  inputText: '',
  isOpen: false,
  itemCount: 0,
  reducer: (state: State, action: Action) => action,
  itemToString: (value: any) => value,
};

export const createState = (): State => ({
  ...initialState,
  id: generateId(),
});
