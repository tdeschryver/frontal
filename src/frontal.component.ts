import {
  Component,
  ContentChild,
  TemplateRef,
  Directive,
  HostListener,
  HostBinding,
  ElementRef,
  Input,
  ContentChildren,
  QueryList,
  Inject,
  forwardRef,
  ChangeDetectionStrategy,
  ChangeDetectorRef,
  OnInit,
  AfterViewInit,
  OnDestroy,
  Output,
  EventEmitter,
  ViewRef,
  ComponentRef,
} from '@angular/core';
import { ControlValueAccessor, NG_VALUE_ACCESSOR } from '@angular/forms';
import { Action, StateChanges } from './actions';
import { State, initialState, createState } from './state';

@Directive({
  selector: '[frontalInput]',
  exportAs: 'frontalInput',
})
export class FrontalInputDirective implements OnInit, AfterViewInit, OnDestroy {
  @HostBinding('attr.role') role = 'combobox';
  @HostBinding('attr.autocomplete') autocomplete = 'off';
  @HostBinding('attr.aria-autocomplete') ariaAutocomplete = 'off';
  @HostBinding('attr.aria-expanded') ariaExpanded = false;
  @HostBinding('attr.id') attrId = `frontal-input-${this.frontal.state.id}`;

  @Input()
  get id() {
    return this.attrId;
  }

  set id(value: any) {
    this.attrId = value;
  }

  constructor(
    @Inject(ElementRef) private element: ElementRef,
    // prettier-ignore
    @Inject(forwardRef(() => FrontalComponent)) private frontal: FrontalComponent, // tslint:disable-line
  ) {}

  ngOnInit() {
    this.frontal.addListener({ id: 'input', listener: this.stateChange.bind(this) });
  }

  ngAfterViewInit() {
    this.setAriaExpanded(this.frontal.state.isOpen);
    this.setValue(this.frontal.state.inputText);
  }

  ngOnDestroy() {
    this.frontal.removeListener('input');
  }

  stateChange(state: State) {
    if (this.ariaExpanded !== state.isOpen) {
      this.setAriaExpanded(state.isOpen);
    }

    if (this.element.nativeElement.value !== state.inputText) {
      this.setValue(state.inputText);
    }
  }

  @HostListener('blur', ['$event'])
  blur(event: Event) {
    this.frontal.inputBlur();
  }

  @HostListener('input', ['$event'])
  input(event: KeyboardEvent) {
    this.frontal.inputChange(event);
  }

  @HostListener('keydown', ['$event'])
  keydown(event: KeyboardEvent) {
    this.frontal.inputKeydown(event);
  }

  setAriaExpanded(value: boolean) {
    this.ariaExpanded = value;
  }

  setValue(value: string) {
    this.element.nativeElement.value = value;
  }
}

@Directive({
  selector: '[frontalLabel]',
  exportAs: 'frontalLabel',
})
export class FrontalLabelDirective {
  @HostBinding('attr.for') attrFor = `frontal-input-${this.frontal.state.id}`;

  @Input()
  get for() {
    return this.attrFor;
  }

  set for(value: any) {
    this.attrFor = value;
  }

  constructor(
    // prettier-ignore
    @Inject(forwardRef(() => FrontalComponent)) private frontal: FrontalComponent, // tslint:disable-line
  ) {}
}

@Directive({
  selector: '[frontalItem]',
  exportAs: 'frontalItem',
})
export class FrontalItemDirective implements OnInit, OnDestroy {
  @Input() value: any;

  constructor(
    // prettier-ignore
    @Inject(forwardRef(() => FrontalComponent)) private frontal: FrontalComponent, // tslint:disable-line
  ) {}

  ngOnInit() {
    this.frontal.addFrontalItem();
  }

  ngOnDestroy() {
    this.frontal.removeFrontalItem();
  }

  @HostListener('mousedown', ['$event'])
  mousedown(event: MouseEvent) {
    this.frontal.itemClick(this);
  }

  @HostListener('mousemove', ['$event'])
  mousemove(event: MouseEvent) {
    this.frontal.itemMove(this);
  }

  @HostListener('mouseleave', ['$event'])
  mouseleave(event: MouseEvent) {
    this.frontal.itemLeave(this);
  }
}

@Directive({
  selector: '[frontalButton]',
  exportAs: 'frontalButton',
})
export class FrontalButtonDirective implements OnInit, AfterViewInit, OnDestroy {
  @HostBinding('attr.type') type = 'button';
  @HostBinding('attr.role') role = 'button';
  @HostBinding('attr.aria-label') ariaLabel = 'close menu';
  @HostBinding('attr.aria-expanded') ariaExpanded = false;
  @HostBinding('attr.aria-haspopup') ariaHasPopup = true;

  constructor(
    // prettier-ignore
    @Inject(forwardRef(() => FrontalComponent)) private frontal: FrontalComponent, // tslint:disable-line
  ) {}

  ngOnInit() {
    this.frontal.addListener({ id: 'button', listener: this.stateChange.bind(this) });
  }

  ngAfterViewInit() {
    this.setAriaExpanded(this.frontal.state.isOpen);
  }

  ngOnDestroy() {
    this.frontal.removeListener('button');
  }

  stateChange(state: State) {
    if (this.ariaExpanded !== state.isOpen) {
      this.setAriaExpanded(state.isOpen);
    }
  }

  @HostListener('click', ['$event'])
  click(event: MouseEvent) {
    if (this.frontal.buttonClick) {
      this.frontal.buttonClick();
    }
  }

  setAriaExpanded(value: boolean) {
    this.ariaExpanded = value;
    this.ariaLabel = value ? 'open menu' : 'close menu';
  }
}

export const FRONTAL_VALUE_ACCESSOR: any = {
  provide: NG_VALUE_ACCESSOR,
  useExisting: forwardRef(() => FrontalComponent), // tslint:disable-line
  multi: true,
};

@Component({
  selector: 'frontal',
  exportAs: 'frontal',
  changeDetection: ChangeDetectionStrategy.OnPush,
  template: `
    <ng-container *ngTemplateOutlet="template; context: state"></ng-container>
  `,
  providers: [FRONTAL_VALUE_ACCESSOR],
})
export class FrontalComponent implements ControlValueAccessor {
  state: State = createState();

  @Input()
  set reducer(fun: (state: State, action: Action) => Action) {
    this.state.reducer = fun;
  }

  @Input()
  set itemToString(fun: (value: any) => string) {
    this.state.itemToString = fun;
  }

  @Input()
  set isOpen(value: boolean) {
    this.state.isOpen = value;
  }

  @Output() change = new EventEmitter<string>();
  @Output() select = new EventEmitter<any>();

  @ContentChild(TemplateRef) template: TemplateRef<any>;
  @ContentChild(FrontalInputDirective) frontalInput: FrontalInputDirective;
  @ContentChildren(FrontalItemDirective) frontalItems: QueryList<FrontalItemDirective>;

  private _stateListeners: { id: string; listener: ((state: State) => void) }[] = [];
  private _onChange = (value: any) => {};
  private _onTouched = () => {};

  constructor(private _changeDetector: ChangeDetectorRef) {}

  writeValue(value: any) {
    const inputText = value ? this.state.itemToString(value) : '';
    this.handle(
      {
        type: StateChanges.InputChange,
        payload: {
          selectedItem: value,
          inputText,
          inputValue: inputText,
        },
      },
      false,
    );
  }

  registerOnChange(fn: any) {
    this._onChange = fn;
  }

  registerOnTouched(fn: any) {
    this._onTouched = fn;
  }

  addListener({ id, listener }: { id: string; listener: (state: State) => void }) {
    this._stateListeners = [...this._stateListeners, { id, listener }];
  }

  removeListener(id: string) {
    this._stateListeners = this._stateListeners.filter(p => p.id !== id);
  }

  toggleMenu() {
    this.handle({
      type: StateChanges.MenuToggle,
      payload: {
        isOpen: !this.state.isOpen,
      },
    });
  }

  openMenu() {
    this.handle({
      type: StateChanges.MenuOpen,
      payload: {
        isOpen: true,
      },
    });
  }

  closeMenu() {
    this.handle({
      type: StateChanges.MenuClose,
      payload: {
        isOpen: false,
      },
    });
  }

  buttonClick() {
    this.handle({
      type: StateChanges.ButtonClick,
      payload: {
        isOpen: !this.state.isOpen,
      },
    });
  }

  inputBlur() {
    const { selectedItem, inputText } = this.getSelected();
    if (this.state.isOpen) {
      this.handle({
        type: StateChanges.InputBlur,
        payload: {
          isOpen: false,
          highlightedIndex: null,
          selectedItem,
          inputText,
          inputValue: inputText,
        },
      });
    }
  }

  inputChange(event: KeyboardEvent) {
    const inputText = (<HTMLInputElement>event.target).value;
    this.handle({
      type: StateChanges.InputChange,
      payload: {
        inputText,
        inputValue: inputText,
        isOpen: true,
        selectedItem: null,
      },
    });
  }

  inputKeydown(event: KeyboardEvent) {
    if (!this.state.isOpen) {
      return;
    }

    const handlers: { [key: string]: () => Action } = {
      ArrowDown: () => {
        event.preventDefault();
        return {
          type: StateChanges.InputKeydownArrowDown,
          payload: {
            selectedItem: null,
            highlightedIndex:
              this.state.itemCount === 0
                ? null
                : ((this.state.highlightedIndex === null ? -1 : this.state.highlightedIndex) + 1) %
                  this.state.itemCount,
          },
        };
      },
      ArrowUp: () => {
        event.preventDefault();
        return {
          type: StateChanges.InputKeydownArrowUp,
          payload: {
            selectedItem: null,
            highlightedIndex:
              this.state.itemCount === 0
                ? null
                : ((this.state.highlightedIndex === null ? 1 : this.state.highlightedIndex) -
                    1 +
                    this.state.itemCount) %
                  this.state.itemCount,
          },
        };
      },
      Enter: () => {
        const { selectedItem, inputText } = this.getSelected();
        return {
          type: StateChanges.InputKeydownEnter,
          payload: {
            isOpen: false,
            highlightedIndex: null,
            selectedItem,
            inputText,
            inputValue: inputText,
          },
        };
      },
      Escape: () => ({
        type: StateChanges.InputKeydownEsc,
        payload: {
          isOpen: false,
          highlightedIndex: null,
          selectedItem: null,
          inputText: '',
          inputValue: '',
        },
      }),
    };

    const handler = handlers[event.key];
    if (handler) {
      this.handle(handler());
    }
  }

  itemClick(item: FrontalItemDirective) {
    const inputText = this.state.itemToString(item.value);
    if (this.state.isOpen) {
      this.handle({
        type: StateChanges.ItemMouseClick,
        payload: {
          isOpen: false,
          highlightedIndex: null,
          selectedItem: item.value,
          inputText,
          inputValue: inputText,
        },
      });
    }
  }

  // MouseMove because we want a user interaction
  // MouseEnter selects an item when the mouse is hovering over an item while typing
  itemMove(item: FrontalItemDirective) {
    if (this.state.isOpen) {
      const index = this.frontalItems.toArray().indexOf(item);
      if (index !== -1 && index !== this.state.highlightedIndex) {
        this.handle({
          type: StateChanges.ItemMouseEnter,
          payload: {
            highlightedIndex: index,
          },
        });
      }
    }
  }

  itemLeave(item: FrontalItemDirective) {
    if (this.state.isOpen) {
      this.handle({
        type: StateChanges.ItemMouseLeave,
        payload: {
          highlightedIndex: null,
        },
      });
    }
  }

  getHighlightedItem() {
    return this.state.highlightedIndex === null
      ? null
      : this.frontalItems.find((_: FrontalItemDirective, index: number) => this.state.highlightedIndex === index);
  }

  handle(action: Action, detactChanges: boolean = true) {
    const { payload } = this.state.reducer(this.state, action);

    const newState = {
      ...this.state,
      ...payload,
    };

    if (newState.selectedItem !== this.state.selectedItem) {
      this._onChange(newState.selectedItem);
      if (newState.selectedItem !== null) {
        this.select.emit(newState.selectedItem);
      }
    }

    if (this.state.inputValue !== newState.inputValue) {
      this.change.emit(newState.inputValue);
    }

    this.state = newState;
    this._stateListeners.forEach(({ listener }) => listener(this.state));

    if (detactChanges) {
      this._changeDetector.detectChanges();
    }
  }

  getSelected() {
    let selectedItem = null;
    const highlighted = this.getHighlightedItem();
    if (highlighted) {
      selectedItem = highlighted.value;
    }
    const inputText = highlighted ? this.state.itemToString(highlighted.value) : '';
    return { selectedItem, inputText };
  }

  addFrontalItem() {
    this.state = {
      ...this.state,
      itemCount: this.state.itemCount + 1,
    };

    this._changeDetector.detectChanges();
  }

  removeFrontalItem() {
    this.state = {
      ...this.state,
      itemCount: this.state.itemCount - 1,
    };

    setTimeout(() => {
      const viewRef = this._changeDetector as ViewRef;
      if (viewRef && !viewRef.destroyed) {
        this._changeDetector.detectChanges();
      }
    });
  }
}
