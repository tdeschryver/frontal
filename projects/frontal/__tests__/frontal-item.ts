import { By } from '@angular/platform-browser';
import { createComponent } from 'ngx-testing-library';
import { FrontalComponent, FrontalItemDirective, StatusMessagePipe, resetId, State, updateState } from '../src';

test('sanity check for attributes', async () => {
  const { items } = await setup();
  expect(Array.from(items)).toMatchSnapshot();
});

test('mouse down selects the item', async () => {
  const { item, frontal, mouseDown } = await setup({
    isOpen: true,
    highlightedIndex: 1,
    highlightedItem: 'foo',
  });

  mouseDown(item);
  expect(frontal.state.value).toMatchObject(
    expect.objectContaining({
      highlightedItem: null,
      highlightedIndex: null,
      selectedItem: 'uno',
      inputText: 'uno',
      inputValue: 'uno',
    }),
  );
});

test('mouse down closes the list', async () => {
  const { item, frontal, mouseDown } = await setup({ isOpen: true });

  mouseDown(item);
  expect(frontal.state.value).toMatchObject(expect.objectContaining({ isOpen: false }));
});

test('mouse move sets the highlighted index', async () => {
  const { item, frontal, mouseMove } = await setup({ isOpen: true });

  mouseMove(item);
  expect(frontal.state.value).toMatchObject(
    expect.objectContaining({
      highlightedItem: 'uno',
      highlightedIndex: 0,
    }),
  );
});

test('mouse leave resets the highlighted index', async () => {
  const { item, frontal, mouseLeave } = await setup({
    isOpen: true,
    highlightedIndex: 0,
    highlightedItem: 'uno',
  });

  mouseLeave(item);
  expect(frontal.state.value).toMatchObject(
    expect.objectContaining({
      highlightedItem: null,
      highlightedIndex: null,
    }),
  );
});

test('highlighted index sets aria selected', async () => {
  const { items, mouseMove } = await setup({
    isOpen: true,
  });

  mouseMove(items[1]);
  expect(items[0].getAttribute('aria-selected')).toBe('false');
  expect(items[1].getAttribute('aria-selected')).toBe('true');

  mouseMove(items[0]);
  expect(items[0].getAttribute('aria-selected')).toBe('true');
  expect(items[1].getAttribute('aria-selected')).toBe('false');
});

async function setup(initialState: Partial<State> = {}) {
  const template = `
    <frontal>
      <ng-template>
        <div frontalItem [value]="'uno'"></div>
        <div frontalItem [value]="'dos'"></div>
        <div frontalItem [value]="'tres'"></div>
      </ng-template>
    </frontal>`;

  resetId();
  const { fixture, mouseDown, mouseLeave, mouseMove, container } = await createComponent(template, {
    declarations: [FrontalComponent, FrontalItemDirective, StatusMessagePipe],
  });

  const frontal = fixture.debugElement.query(By.css('frontal')).componentInstance as FrontalComponent;

  if (initialState) {
    frontal.actions.next(updateState(initialState));
  }

  fixture.changeDetectorRef.detectChanges();
  await fixture.whenStable();

  return {
    fixture,
    frontal,
    item: container.querySelector('[frontalItem]') as HTMLElement,
    items: container.querySelectorAll('[frontalItem]') as NodeListOf<HTMLElement>,
    mouseDown,
    mouseLeave,
    mouseMove,
  };
}
