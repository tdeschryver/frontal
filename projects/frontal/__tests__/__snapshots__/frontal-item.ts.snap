// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sanity check for attributes 1`] = `
Array [
  <div
    aria-selected="false"
    frontalitem=""
    id="frontal-item-0-1"
    ng-reflect-value="uno"
    role="option"
  />,
  <div
    aria-selected="false"
    frontalitem=""
    id="frontal-item-0-2"
    ng-reflect-value="dos"
    role="option"
  />,
  <div
    aria-selected="false"
    frontalitem=""
    id="frontal-item-0-3"
    ng-reflect-value="tres"
    role="option"
  />,
]
`;
